designs/helm-chart-autogenerate-plugin.md

```diff
@@ -0,0 +1,627 @@
+
+- **Method 1:Let Helm Do It For You:**  Place CRDs in the `crds/` directory. Helm installs these CRDs during the initial
+  install but does not manage upgrades or deletions.
+- **Method 2:Separate Charts:**  Place the CRD definition in one chart
+  and the resources using the CRD in another chart.
+  This method requires separate installations for each chart.
+
+**Raised Considerations and Concerns**
+- **Use Helm crd directory** The upgraded chart versions will silently ignore CRDs even if they differ from the installed versions. This could lead to surprising and unexpected behavior. Therefore, Kubebuilder should not encourage or promote this approach.
+- **Templates Folder**: Moving CRDs to the `templates` folder facilitates upgrades but uninstalls CRDs when the operator is uninstalled. However, it allows users easier manage the CRDs and install them on upgrades. It is a common approach adopted by maintainers but is not considered a good practice by Helm itself.
+- **Separate Helm Chart for CRDs:** This approach allows control over both CRD and operator versions without deleting CRDs when the operator chart is deleted. Also, follows the HelmChart best practices. Another problem with this approach is to ensure that the CRDs will be applied before the CRs since both will be under the template directory.
+- **When Webhooks are used:** If a CRD specifies, for example, a conversion webhook, the "API chart" needs to contain the CRDs and the webhook `service/workload`.
+  It would also make sense to include `validating/mutating` webhooks, requiring the scaffolding of separate main modules and image builds for
+  webhooks and controllers which does not shows to be compatible with Kubebuilder Golang scaffold.
+
+**Proposed Solution**
+
+Follow the same approach adopted by [Cert-Manager](https://cert-manager.io/docs/installation/helm/).
+Add the CRDs under the `template` directory and have a spec in the `values.yaml`
+which will define if the CRDs should or not be applied
+
+```shell
+helm install|upgrade \
+  myrelease \
+  --namespace my-namespace \
+  --set `crds.enabled=true`
+```
+
+Also, add another spec to the `values.yaml` to not allow the CRDs
+be deleted when the helm is uninstalled:
+
+```yaml
+  # START annotations {{- if .Values.crds.keep }}
+  annotations:
+    helm.sh/resource-policy: keep
+  # END annotations {{- end }}
+```
+
+By last, to ensure that the CRD will be installed before
+the samples under the template we will need to use helm-hooks:
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+metadata:
+  name: myresources.example.com
+  annotations:
+    "helm.sh/hook": pre-install, pre-upgrade
+    "helm.sh/hook-weight": "0"
+```
+
+Additionally, we want to
+scaffold separate charts for the APIs and support both.
+An example of this approach provided as feedback
+was [karpenter-provider-aws](https://github.com/aws/karpenter-provider-aws/tree/main/charts).
```

Comment:
@joelanford @Jay-Madden @FourFifthsCode ^ see the proposed solution
The idea would have both and let maintainers/consumers of the project decide.

WDYT?

Or do you think we should only do the separate chart for crds since that follow the good practices?